File: ios_app/GenAIAcademyDashboard/ViewModels/LoginViewModel.swift
//
//  LoginViewModel.swift
//  GenAIAcademyDashboard
//
//  ViewModel for login screen
//

import Foundation
import SwiftUI

@MainActor
class LoginViewModel: ObservableObject {
    @Published var email = ""
    @Published var password = ""
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let authService = AuthService.shared
    
    /// Validate email format
    func isValidEmail() -> Bool {
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        return emailPredicate.evaluate(with: email)
    }
    
    /// Validate form before submission
    func validateForm() -> Bool {
        if email.isEmpty {
            errorMessage = "Email is required"
            return false
        }
        
        if !isValidEmail() {
            errorMessage = "Please enter a valid email address"
            return false
        }
        
        if password.isEmpty {
            errorMessage = "Password is required"
            return false
        }
        
        if password.count < 3 {
            errorMessage = "Password must be at least 3 characters"
            return false
        }
        
        return true
    }
    
    /// Perform login
    func login() async {
        // Clear previous errors
        errorMessage = nil
        
        // Validate form
        guard validateForm() else {
            return
        }
        
        isLoading = true
        
        do {
            try await authService.login(email: email, password: password)
            // Success - authentication state will be updated by AuthService
        } catch let error as NetworkError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "An unexpected error occurred. Please try again."
        }
        
        isLoading = false
    }
}

File: flutter_app/lib/viewmodels/login_viewmodel.dart
import 'package:flutter/foundation.dart';
import '../services/auth_service.dart';

/// ViewModel for login screen (MVVM pattern)
class LoginViewModel extends ChangeNotifier {
  final _authService = AuthService();
  
  bool _isLoading = false;
  String? _errorMessage;
  bool _isAuthenticated = false;
  
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;
  bool get isAuthenticated => _isAuthenticated;
  
  /// Validate email format
  bool isValidEmail(String email) {
    return RegExp(r'^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$').hasMatch(email);
  }
  
  /// Validate password (minimum 6 characters)
  bool isValidPassword(String password) {
    return password.length >= 6;
  }
  
  /// Login with email and password
  Future<bool> login(String email, String password) async {
    // Validation
    if (!isValidEmail(email)) {
      _errorMessage = 'Please enter a valid email address';
      notifyListeners();
      return false;
    }
    
    if (!isValidPassword(password)) {
      _errorMessage = 'Password must be at least 6 characters';
      notifyListeners();
      return false;
    }
    
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();
    
    try {
      await _authService.login(email, password);
      _isAuthenticated = true;
      _isLoading = false;
      _errorMessage = null;
      notifyListeners();
      return true;
    } catch (e) {
      _isLoading = false;
      _errorMessage = e.toString().replaceAll('Exception: ', '');
      _isAuthenticated = false;
      notifyListeners();
      return false;
    }
  }
  
  /// Clear error message
  void clearError() {
    _errorMessage = null;
    notifyListeners();
  }
}
